### PR TITLE
Add post-v4.0 freq VDS split fix to ensure all hom refs are kept in all samples

### PR DESCRIPTION
**Pull request by mike-w-wilson**
This fixes the existing 4.0 code so all hom ref calls are kept in the subsets after the split and densify. I tried to document extensively to make sure future users knew this wasnt used in v4.0 but would have the desired  freq outcome if it were.

### CODE REVIEW

File: gnomad_qc/v4/annotations/generate_freq.py [hunk 1]
```diff
@@ -1017,7 +1048,24 @@ def main(args):
...
+            # removing any variant row that does not have a defined entry. When the
+            # VDS is densified, this results in a loss of all homozygous reference calls
+            # in the reference data sparse MT at these dead allele sites. Because we
+            # had to split the VDS for v4.0 with intentions of rejoining the two subset
+            # VDSs across all variant sites, this unexpected behavior resulted in the
+            # loss of all homozygous reference calls in the reference data sparse MT and
+            # thus reduced Allele Number (AN) for any bi-allelic variant exclusive to
+            # one of the subsetted VDSs. The replacement code below uses our own custom
+            # function to split the VDS and keep the dead alleles in the reference data.
+            # It was never run in production, but it is the correct way to split the VDS
```
Comment (KoalaQin) [inline gnomad_qc/v4/annotations/generate_freq.py:1066]: So we didn't rerun this but we have separate production codes to correct freq and to get all sites AN? Do you think it's worth noting the fix here? I'm looking at this PR https://github.com/broadinstitute/gnomad_qc/pull/556 for fix_freq_an.py.

File: gnomad_qc/v4/annotations/generate_freq.py [hunk 2]
```diff
@@ -1017,7 +1048,24 @@ def main(args):
                 "Splitting VDS by ukb_sample annotation to reduce data size for"
                 " densification..."
             )
-            vds_dict = split_vds_by_strata(vds, strata_expr=vds.variant_data.ukb_sample)
+            # For v4.0 the line of code this comment block was:
+            # vds_dict = split_vds_by_strata(vds, strata_expr=vds.variant_data.ukb_sample)
+            #
+            # This uses hail's hl.vds.filter_samples() function and keeps the default
```
[the remark this answers is not in the document]
Comment (KoalaQin) [reply gnomad_qc/v4/annotations/generate_freq.py:1054]: Is Hail Team correcting this unexpected behavior in future release? If so, we might have to put the versions of Hails that were impacted.

**Review by KoalaQin — CHANGES_REQUESTED** [review]
I have a few comments (or questions) since I didn't follow what we've done for the fix.

File: gnomad_qc/v4/annotations/generate_freq.py (hunk 1 above)
Comment (mike-w-wilson) [reply gnomad_qc/v4/annotations/generate_freq.py:1066]: Yeah good point, I'll reference the script and ticket.

File: gnomad_qc/v4/annotations/generate_freq.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mike-w-wilson) [reply gnomad_qc/v4/annotations/generate_freq.py:1054]: For your first question, it depends if you plan on joining the vds back together after using hl.vds.filter_samples. If you do, then yes its a problem. This impacts all versions of hail that have hl.vds.filter_samples() so we can note that its true through 0.2.128 here.

**Comment by mike-w-wilson** [comment]
Thakns @KoalaQin , back to you!

**Review by KoalaQin — APPROVED** [review]
LGTM! 
Remember to set a ticket to modify `split_vds_by_strata` in gnomad_methods.

**Comment by mike-w-wilson** [comment]
https://github.com/broadinstitute/gnomad_methods/issues/683

**Pull request merged**